### PR TITLE
refactor(ActionDescriptor,ActionList,ActionReference): improve type communication

### PR DESCRIPTION
**Pull request by TsXor**
There is a problem in the current AM part of this api:
When you use "get" functions, what you get **is** AM thing, but **not** an AM thing as well.
It **is** an AM thing: you can use those "put" and "get" functions documented in [Adobe's js scripting reference](https://raw.githubusercontent.com/Adobe-CEP/CEP-Resources/master/Documentation/Product%20specific%20Documentation/Photoshop%20Scripting/photoshop_scriptref_js-cc2014.pdf)
It is **not** an AM thing: It is just a COM bind object, using 'getType' will return number.
So I made this modification to unify.

Also, it seems that the functions written for the AM classes are not nessesary because it seems that COM bindings have already done that.

### CODE REVIEW

**Comment by loonghao** [comment]
@TsXor 
Thank you for your contribution

You can use the following command line to format codes.

setting up the dev environment
```
pip install poetry
poetry install
```
format code
```
poetry run black photoshop
poetry run isort photoshop
```
commit code

```
git add photoshop/api/action_descriptor.py
git add photoshop/api/action_list.py
git add photoshop/api/action_list.py
git add photoshop/api/action_reference.py
```
https://github.com/commitizen-tools/commitizen
```
cz commit 
```

**Comment by TsXor** [comment]
> @TsXor Thank you for your contribution
> 
> You can use the following command line to format codes.
> 
> setting up the dev environment
> 
> ```
> pip install poetry
> poetry install
> ```
> 
> format code
> 
> ```
> poetry run black photoshop
> poetry run isort photoshop
> ```
> 
> commit code
> 
> ```
> git add photoshop/api/action_descriptor.py
> git add photoshop/api/action_list.py
> git add photoshop/api/action_list.py
> git add photoshop/api/action_reference.py
> ```
> 
> https://github.com/commitizen-tools/commitizen
> 
> ```
> cz commit 
> ```

Thanks, in fact I just test the code locally and manually sync them to github in browser. I'll try these.

**Comment by TsXor** [comment]
I thought I should patch app.executeAction() and app.executeActionGet(), but it just works fine! Maybe that's the magic of duck typing.

**Comment by TsXor** [comment]
@loonghao Is there an api to find out which version of photoshop is running?

**Comment by loonghao** [comment]
@TsXor 

Three things need to be addressed before merging this MR

-  Change the name of the files:

    ```
    photoshop/api/_AMtypebinder.py >> photoshop/api/_action_manager_typebinder.py
    ```
    ```
    photoshop/api/_AMunify.py >> photoshop/api/_action_manager_unify.py
    ```
    reference:
    https://google.github.io/styleguide/pyguide.html#316-naming

- We need to add some examples for the current new functions

  like these:
  https://github.com/loonghao/photoshop-python-api/tree/main/examples


- this commit tag should be `feat` because you added some new functions
image

Thanks for taking the time to contribute the code.

**Comment by TsXor** [comment]
>     * We need to add some examples for the current new functions
>       like these:
>       https://github.com/loonghao/photoshop-python-api/tree/main/examples
> 
>     * this commit tag should be `feat` because you added some new functions
>       image
> 
> 
> Thanks for taking the time to contribute the code.

I think these 2 things are not needed, becauses it does not change anything when using this api. It just makes sure ActionDescriptor always have ActionDescriptor as its type (the same for ActionList, ActionReference). Files for typebinder and unify have a `_` before their name to show that they are "internal" -- no one should use them unless they know what they are doing.

These 2 pictures may show why to make typebinder and unify:
typebinder

**Comment by TsXor** [comment]
@loonghao I think you should know about the implementation of unify well, because _core.py uses the same thing:
```python
    def __getattribute__(self, item):
        try:
            return super().__getattribute__(item)
        except AttributeError:
            return getattr(self.app, item)
```
Well, that equals to:
```python
    def __getattr__(self, item):
        try:
            return getattr(self.app, item)
        except AttributeError:
            raise AttributeError
```
So I have now found out that unify is not needed. `_core` is almost the same thing as unify. Just monkey patching the `app` attribute is enough.

**Comment by TsXor** [comment]
So I also found that lots of the code in this repo is to fix the drawback of `_core`, exactly the same as unify.
We can't exactly find which attribute is overwritten in subclass and which attribute is inherited from `object` class.
I think this repo would greatly shrink if we fix this problem.
venn

**Comment by loonghao** [comment]
@TsXor 
COM itself is property dynamic binding, but some functions are inconsistent with the description in the JS documentation, so we made some patches,
Then the re-implementation of these functions is not only for patching but also for the convenience of code completion and improving ease of use.

File: photoshop/api/action_descriptor.py [hunk 1]
```diff
@@ -8,204 +8,127 @@
...
-        """Sets the value for a key whose type is integer."""
         self.app.putInteger(key, value)
 
     def putLargeInteger(self, key: int, value: int):
-        """Sets the value for a key whose type is large integer."""
         self.app.putLargeInteger(key, value)
 
-    def putList(self, key: int, value: ActionList):
-        """Sets the value for a key whose type is an ActionList object."""
+    def putList(self, key: int, value: "ActionList"):
```
[the remark this answers is not in the document]
Comment (TsXor) [reply photoshop/api/action_descriptor.py:118]: It is defined in _action_manager_typebinder, and a subclass will inherit this method there.

File: photoshop/api/action_list.py [hunk 2]
```diff
@@ -4,54 +4,122 @@
...
+    def putInteger(self, value: int):
+        self.app.putInteger(value)
+
+    def putLargeInteger(self, value: int):
+        self.app.putLargeInteger(value)
+
+    def putList(self, value: "ActionList"):
+        self.app.putList(value)
+
+    def putObject(self, classID: int, value: "ActionDescriptor"):
```
[the remark this answers is not in the document]
Comment (TsXor) [reply photoshop/api/action_list.py:112]: It is defined in _action_manager_typebinder, and a subclass will inherit this method there.

File: photoshop/api/action_list.py [hunk 3]
```diff
@@ -4,54 +4,122 @@
...
+        self.app.putObject(classID, value)
+
+    def putPath(self, value: pathlib.Path):
+        self.app.putPath(str(value))
+
+    def putReference(self, value: "ActionReference"):
```
[the remark this answers is not in the document]
Comment (TsXor) [reply photoshop/api/action_list.py:118]: It is defined in _action_manager_typebinder, and a subclass will inherit this method there.

File: photoshop/api/action_descriptor.py [hunk 4]
```diff
@@ -8,204 +8,127 @@
...
 
-    def putPath(self, key: int, value: str):
-        """Sets the value for a key whose type is path."""
-        self.app.putPath(key, value)
+    def putPath(self, key: int, value: pathlib.Path):
+        self.app.putPath(key, str(value))
 
-    def putReference(self, key: int, value: ActionReference):
-        """Sets the value for a key whose type is an object reference."""
+    def putReference(self, key: int, value: "ActionReference"):
```
[the remark this answers is not in the document]
Comment (TsXor) [reply photoshop/api/action_descriptor.py:127]: It is defined in _action_manager_typebinder, and a subclass will inherit this method there.

**Comment by TsXor** [comment]
Well, in order to get rid of imcompatibilities and test getData() function, I am moving to CC 2014, and I'm going to "copy the doc into code", that may take a while.
Hope we can decode the black magic of AM soon. As Hilbert once said, "We must know, we will know."

**Comment by TsXor** [comment]
也就是说其实在Session里用了AM代码的例子全是扯淡🤣

Session的ActionDescriptor, ActionList, ActionReference全是一个对象，这就是为什么用了Session的例子里这仨类都不实例化。
那为啥能正常工作呢？因为正如我一个issue中所描述，AM对象有个特性，子对象的更改都不会同步给父对象，一个AM对象嵌到其他AM对象里就像是copy了再存一样，这也就意味着你一直在Session里用一个对象还真有可能正常工作。好家伙，连上了，全连上了！🤣负负得正了属于是

但是也有几率翻车：你所有AM对象都用的一个，就像用同一个烧瓶装完试剂A，然后直接往里倒试剂B，迟早会有那么几次发生化学反应。况且这样导致了在Session模式和非Session模式中行为不一致，因此我强烈建议改了它。

**Pull request closed without merging**